homeassistant/components/viam/config_flow.py

```diff
@@ -0,0 +1,185 @@
+        secret = data[CONF_SECRET]
+
+    creds = Credentials(type=credential_type, payload=secret)
+    opts = DialOptions(auth_entity=auth_entity, credentials=creds)
+    client = await ViamClient.create_from_dial_options(opts)
+
+    # If you cannot connect:
+    # throw CannotConnect
+    if client:
+        locations = await client.app_client.list_locations()
+        location = await client.app_client.get_location(next(iter(locations)).id)
+
+        # Return info that you want to store in the config entry.
+        return (location.name, client)
+
+    raise CannotConnect
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for viam."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize."""
+        self._title = ""
+        self._client: ViamClient | None = None
+        self._data: dict[str, Any] = {}
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self._data.update(user_input)
+            return await self.async_step_auth()
+
+        return self.async_show_form(
+            step_id="user", data_schema=STEP_AUTH_USER_DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_auth(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                self._data.update(user_input)
+                (title, client) = await validate_input(self._data)
+                self._title = title
+                self._client = client
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return await self.async_step_robot()
+
+        schema = STEP_AUTH_ROBOT_DATA_SCHEMA
+        if self._data.get(CONF_CREDENTIAL_TYPE) == CRED_TYPE_API_KEY:
+            schema = STEP_AUTH_ORG_DATA_SCHEMA
+
+        return self.async_show_form(
+            step_id="auth",
+            data_schema=schema,
+            errors=errors,
+        )
+
+    async def async_step_robot(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Select robot from location."""
+
+        app_client = self._get_app_client()
+        locations = await app_client.list_locations()
+        robots = await app_client.list_robots(next(iter(locations)).id)
+        if user_input is not None:
+            robot_id = next(
+                robot.id for robot in robots if robot.name == user_input[CONF_ROBOT]
+            )
+            self._data.update({CONF_ROBOT_ID: robot_id})
+            self._close_client()
```

Comment:
The client will only be closed on success and not when the user aborts the config flow

Reply:
This will be called: https://github.com/home-assistant/core/blob/4c6767056668b039d99bd5dc5a215dfd727726b7/homeassistant/data_entry_flow.py#L855-L857

Here's an example of how it can be used: https://github.com/home-assistant/core/blob/4c6767056668b039d99bd5dc5a215dfd727726b7/homeassistant/components/improv_ble/config_flow.py#L406-L409

Reply:
Thanks @emontnemery! That was very helpful.